README.md
CAN bus overhead, stuff bits and effective bitrate calculator
===============================================================================

### [Click to use it in your browser!](https://nik89.github.io/CanOverhead/)

The [CAN bus](https://en.wikipedia.org/wiki/CAN_bus)
is a robust communication bus and protocol used in vehicle,
allowing frames with 11 or 29-bit identifiers in the header and payload sizes
of up to 8 bytes for classic CAN and up to 64 for the
[CAN FD](https://en.m.wikipedia.org/wiki/CAN_FD) variant.

This JavaScript in-browser calculator helps you compute:

- the overhead of a frame in space: size of header and trailer
- the stuff bits: these are bits injected after each streak of 5 bits
- the effective bitrate: compute the transfer speed and time of your data taking
  into account also the time it takes to transfer the frame overhead.


Usage
---------------------------------------

### Online

Head over to the
[**website of the calculator**](https://nik89.github.io/CanOverhead/)
to use it. Simple as that!


### Locally

You can also directly download the page and use it in any modern browser,
also offline. **Right-click** on the
[website link](https://nik89.github.io/CanOverhead/)
and choose _"Save as..."_ to download the `index.html` page. Rename it
as you please.

We also make the same file available in the
[releases on GitHub](https://github.com/Nik89/CanOverhead/releases).



Privacy
---------------------------------------

The calculator has no backend server, works in your browser locally,
no data is transferred anywhere. If you are paranoid,
you can also use it after setting your browser to stay offline (in Firefox:
_File > Work Offline_) after loading the page and the calculator still operates
- completely disconnecting the computer from the internet also works.

Also: it does not use any cookies.



Compilation
---------------------------------------

The calculator is a set of HTML+CSS+JS files, so it does not need actual
compilation to work. You can simply download or clone the repository and
open the `index.html` file in your favourite browser for it to work
(except some URLs to license, changelog etc. may be broken).

To minify and merge all HTML, CSS and JS files into a single `index.html`
the Python3 script `minify.py` must be run. This will generate the compact
all-in-one `index.html` in the `minified` directory.
You will need some dependencies to run the script, so be sure to run
the following:
```
pip install -r requirements.txt
```
We recommend using a
[virtual environment](https://docs.python.org/3/tutorial/venv.html).


CHANGELOG.md
Changelog
===============================================================================

All notable changes to this project will be documented in this file.

The format is based on
[Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

*******************************************************************************

[1.2.0] - 2021-04-03
----------------------------------------

Bitrates, transfer time and overheads.


### Added

- Bitrates list available: 1 Mbit/s, 800, 500, 250, 125, 50, 20, 10, 5 kbit/s.
- Calculation of gross transfer time.
- Calculation of overhead percentage and effective bitrate.
- Add question mark icon with popup for extra info about the fields on mouseover
  for some input and some output fields.


### Fixed

- Indentation of nested lists in Markdown files was not maintained after
  conversion to HTML.
- Added missing title to v1.1.0 release in this Changelog.
- Fix minified directory to be "docs" instead of "minified" so it can be
  deployed with GitHub pages easily.



[1.1.0] - 2021-04-03
----------------------------------------

Added 29 bits CAN ID, RTR frames, highlight stuff bits, and complete graphical
overhaul.


### Added

- Support for extended frames (29-bit CAN ID) and Remote Transmission Request
  (RTR frames). The frame type can be selected as first field of the form,
  impacting the input form itself (to ask for payload or DLC field)
  and obviously the resulting output.
- Dropdown menu to select the base of the CAN ID to avoid the user typing
  "0x" or "0b".  Defaults to hex. If the user also types the prefix "0x" or "0b"
  and it matches with the selected base, then the prefix is ignored.
- Run computation when pressing Enter keyboard key.
- Highlight incorrect input fields with a red border when the computation is
  run.
- Highlight stuff bits in the output: underlined and with a different color.
- Minifying instructions and privacy notice included in Readme file.


### Changed

- Improved style of the following parts into a more modern look-and-feel
  - input form
  - computing button
  - whole-frame bit sequence section
  - frame fields, now provided as a table. Includes alternating row colors,
    mouseover effects and hex format column
- More verbose and detailed error messages for the various incorrect inputs.
- Improved description of input form's payload field, clarifying what is ignored
  in the user input.
- Autofocus on CAN ID input field.
- Input text fields contain examples for the user input.
- Renamed _Pause-after-frame_ field to _Inter-Frame Space_ for clarity
  and to match with other descriptions of the CAN frames found online.
- Moved error messages to one location underneath the input form.
- Deployment script merges HTML, CSS and JS files into one HTML file, making
  it much easier to deploy. Removed deployment into `gh-pages` branch, as
  now it can be done directly from the `main` branch, `minified` sub-folder.


### Fixed

- Broken links in readme to the deployed web page (missing "/" at URL end).
- Display error msg in case of empty CAN ID instead of assuming the ID == 0.
- Handle empty or whitespace-only payloads as valid input (no data).
- Footer's too small bottom margin.
- Typo in footer disclaimer.
- noscript-message about JavaScript being required is not all-h2 anymore.
  Additional styling is added for it to make it at least slightly decent
  looking.
- Display a generic error message in case of unknown and uncaught errors
  at any level in the execution of the code instead of silently doing nothing
  and crashing the app.
- Fixed non-hex payload chars being parsed into a 0 value.
- Some fields of the CAN frame did not have space separators every 4 bits.
- Fully reset the input form, clearing its content on page reload/refresh.



[1.0.0] - 2021-01-31
----------------------------------------

First version, minimum sets of features.
Classic CAN 11-bit ID supported, listing all fields, whole frame as bits pre
and post-stuffing, max worst-case theoretical frame length.

### Added

- Classic CAN with 11-bit IDs with the CAN ID and payload content known:
  - CAN ID input accepted in binary, hex and decimal format, e.g. `0b0010010`,
    `0xAA`, `42`.
  - Payload content input accepted in hex format, e.g. `AA, 0x0C...`.
  - Compute and display the whole frame as bits (header, payload, trailer)
    with format `0110 1011 1001 1000 ...` without stuff bits.
    The computation of the
    [CRC-15](https://www.can-cia.org/can-knowledge/can/crc/)
    is included in the computation of the trailer.
  - Same as previous point but with stuff bits included.
  - Compute and display the exact amount of added stuff bits.
  - Compute the size in bits of the whole frame (header, payload, trailer)
    without stuff bits.
  - Compute the size range in bits of the whole frame when taking also
    stuff bits into account (min = no stuff bits applied, max = max stuff
    bits applied for CAN ID and payload).
  - Deployment script that prepares and minifies all the files and compresses
    them in `.gz` format


deploy.py
#!/usr/bin/env python3
# coding: utf-8
#
# Published under the terms of the BSD 3-Clause license. See LICENSE.md for
# details.

"""
Deployment script to prepare the web pages of CanOverhead project
to be published on the GitHub pages.

What it does:
1. Creates a "builds/" directory next to this file, deleting any previous one.
   Our ready-for-production web pages will end up here.
2. Converts markdown files to HTML. No CSS is being added, but it's still
   nicer than reading txt files in the browser. Outputs the html files
   in the build directory.
3. Minifies the HTML, CSS and JS files, outputting them into the build
   directory.
4. Inlines the CSS and JS files into the HTML file to have a single output
   file, easier to deploy and/or send around (e.g. email attachment).
   It's also simpler to download from a static server.
"""

import os
import re
import shutil

import markdown
import requests

THIS_FILE_ABS_DIR = os.path.dirname(__file__)
# Force working dir to be the one where this file is,
# allowing us to create the output directory next to this file instead of
# where the file is called from.
os.chdir(THIS_FILE_ABS_DIR)
BUILD_DIR = os.path.join(THIS_FILE_ABS_DIR, "builds")


def cleanup():
    """Removes all files in the build directory, creating it if necessary"""
    shutil.rmtree(BUILD_DIR, ignore_errors=True)
    os.mkdir(BUILD_DIR)


def md2html(input_file_name: str):
    """
    Converts a given Markdown file into an HTML file into the build directory
    Args:
        input_file_name: file name of the MD input file
    """
    output_file_name = input_file_name.lower().rsplit(".", maxsplit=1)[0]
    output_file_name += ".html"
    output_file_name = os.path.join(BUILD_DIR, output_file_name)
    markdown.markdownFromFile(input=input_file_name, output=output_file_name,
                              encoding="UTF-8", output_format="html")


def prepend_comment_in_index_file():
    """Prepend some comment lines in index.html file"""
    filename = os.path.join(BUILD_DIR, "index.html")
    line = ("<!-- Oh hi there! If you want to see the "
            "non-minified source code, check\n"
            "the project GitHub repository: "
            "https://github.com/Nik89/CanOverhead -->\n")
    with open(filename, 'r+', encoding="UTF-8") as file:
        content = file.read()
        file.seek(0)
        file.write(line + content)


def minify_html(input_file_name: str):
    """
    Minify a given HTML file into the build directory
    Args:
        input_file_name: file name of the HTML input file
    """
    _minify(input_file_name, 'https://html-minifier.com/raw')


def minify_css(input_file_name: str):
    """
    Minify a given CSS file into the build directory
    Args:
        input_file_name: file name of the CSS input file
    """
    _minify(input_file_name, 'https://cssminifier.com/raw')


def minify_js(input_file_name: str):
    """
    Minify a given JS file into the build directory
    Args:
        input_file_name: file name of the JS input file
    """
    _minify(input_file_name, 'https://javascript-minifier.com/raw')


def _minify(input_file_name: str, web_api_url: str) -> None:
    """
    Minify a given file into the build directory using the given
    web API.
    Args:
        input_file_name: file name of the JS input file
        web_api_url: URL of the compression web API
    """
    output_file_name = os.path.join(BUILD_DIR, input_file_name)
    file_content = open(input_file_name, mode='rb').read()
    response = requests.post(web_api_url, data={'input': file_content})
    minified = response.text
    if minified.startswith('// Error'):
        raise RuntimeError(
            'File cannot be compressed correctly:\n' + minified)
    open(output_file_name, "w", encoding="UTF-8").write(minified)


def inline_js_and_css_into_html(html_file_name: str) -> None:
    """Inline the .css files found in the <link> tags
    and inline the .js files found in <script> tags
    into the html file in-place to combine them into a single file."""
    html_file_name = os.path.join(BUILD_DIR, html_file_name)
    html = open(html_file_name).readlines()
    inlined = []
    for line in html:
        if '<link' in line and 'stylesheet' in line:
            css_file = re.search(r'[a-zA-Z0-9_.]+\.css', line)
            if css_file is None:
                raise RuntimeError("Can't extract CSS file name from HTML.")
            css_file = os.path.join(BUILD_DIR, css_file.group(0))
            inlined.append('<style>' + open(css_file).read() + '</style>\n')
            os.remove(css_file)
        elif '<script' in line:
            js_file = re.search(r'[a-zA-Z0-9_.]+\.js', line)
            if js_file is None:
                raise RuntimeError("Can't extract JS file name from HTML.")
            js_file = os.path.join(BUILD_DIR, js_file.group(0))
            inlined.append('<script>' + open(js_file).read() + '</script>\n')
            os.remove(js_file)
        else:
            inlined.append(line)
    open(html_file_name, 'w').writelines(inlined)


def build():
    """Converts the Markdown files in the repo to html, minifies the html,
    css and js files, inlines them into a single html file and provides all
    of the results in the builds directory, ready to be deployed."""
    cleanup()
    md2html("CHANGELOG.md")
    md2html("LICENSE.md")
    md2html("README.md")
    md2html("ROADMAP.md")
    # Note: minifying first, then inlining, as it's much easier to scan
    # a minified index.html file; additionally the HTML minifier does not
    # properly minify inlined CSS and JS, so we minify them separately
    # and then perform the inlining.
    minify_html("index.html")
    minify_css("style.css")
    minify_js("CanOverhead.js")
    minify_js("HtmlToLibAdapter.js")
    minify_js("TestCanOverhead.js")
    inline_js_and_css_into_html("index.html")
    prepend_comment_in_index_file()


if __name__ == "__main__":
    build()


scripts.js
let computeV1 = () => {
    // Parse input fields
    let payloadLengthByte = parseInt(
        document.getElementById("v1-in-payload-length").value);

    if (isNaN(payloadLengthByte)
        || payloadLengthByte < 0
        || payloadLengthByte > 8) {// TODO magic numbers
        // Set error and clear existing output
        document.getElementById("v1-input-error").innerHTML = "The value should be between 0 and 8 bytes";
        document.getElementById("v1-out-frames-bits-min").innerHTML = "";
        document.getElementById("v1-out-frames-bits-max").innerHTML = "";
    } else {
        // Clear errors
        document.getElementById("v1-input-error").innerHTML = "";

        // Calculate
        let payloadLengthBit = payloadLengthByte * 8;
        let payloadLengthBitMax = BitSequence.maxLengthAfterStuffing(payloadLengthBit);

        // Set outputs
        document.getElementById("v1-out-frames-bits-min").innerHTML = payloadLengthBit;
        document.getElementById("v1-out-frames-bits-max").innerHTML = payloadLengthBitMax;
    }
}
document.getElementById("v2-in-payload").value = "1111111111000000000011111111110000000000111111111100000000001111";
let computeV2 = () => {
    // TODO inputs reading, cleanup and conversion
    // ID in hex format as a number. Examples:
    // - 1F
    // - 0x1F
    // - 0X1F
    // - 0x1f
    // - f
    // - F
    // - 0xF
    // Idea: strip "0x"/"0X", uppercase, convert to Number
    //
    // Payload in hex format as a hex string. Examples:
    // - AA
    // - 0xAA
    // - 0Xaa
    // - A
    // - AA, BB, CC
    // - 0xAA, 0xBB, 0xCC
    // (As above)
    // Idea: strip same as above, convert to Uint8Array

    // TODO pass inputs to CanOverhead.js::CanFrame11Bit
    // TODO print CanFrame11Bit method results in pretty manner

    // Parse input fields
    let identifierStr = document.getElementById("v2-in-identified").value;
    identifierStr = identifierStr.padStart(11, "0");
    let payloadStr = document.getElementById("v2-in-payload").value;
    payloadStr = payloadStr.padStart(nearestLargerMultipleOf8(payloadStr.length), "0");

    if (identifierStr.length > 11) {
        // Set error and clear existing output
        document.getElementById("v2-input-id-error").innerHTML = "The identifier should be of 11 bits at most";
        //document.getElementById("").innerHTML = "";
        //document.getElementById("").innerHTML = "";
    } else if (payloadStr.length > 64 || (payloadStr.length % 8 !== 0)) {
        // Set error and clear existing output
        document.getElementById("v2-input-pl-error").innerHTML = "Payload should be less than 64 bits and should be N bites long";
    } else {
        // Clear errors
        document.getElementById("v2-input-id-error").innerHTML = "";
        document.getElementById("v2-input-pl-error").innerHTML = "";

        // Calculate
        let payloadLengthBit = payloadLengthByte * 8;
        let payloadLengthBitMax = BitSequence.maxLengthAfterStuffing(payloadLengthBit);

        // Set outputs
        document.getElementById("v1-out-frames-bits-min").innerHTML = payloadLengthBit;
        document.getElementById("v1-out-frames-bits-max").innerHTML = payloadLengthBitMax;
    }
}


function nearestLargerMultipleOf8(x) {
    return Math.floor((x + 7) / 8) * 8;
}


TestCanOverhead.js
/**
 * @file Unit tests of the CanOverhead.js library using simple console
 * assertions that point to the failing line.
 * @licence BSD 3-clause license. See LICENSE.md for details.
 */

let check = console.assert; // Shorter-name alias
/**
 * Compares two arrays for equality of type, length and content.
 * @param {Array|Uint8Array} a first array
 * @param {Array|Uint8Array} b second array
 * @returns {boolean} true if they are equal, false otherwise
 */
function arrayEqual(a, b) {
    let bothArraysOfSameSize = (Array.isArray(a) && Array.isArray(b) && a.length === b.length);
    let bothUint8ArraysOfSameSize = (a instanceof Uint8Array && b instanceof Uint8Array && a.length === b.length);
    if (bothArraysOfSameSize || bothUint8ArraysOfSameSize) {
        // Compare them field by field
        for (let i = 0; i < a.length; i++) {
            if (a[i] !== b[i]) return false;
        }
        return true;
    } else {
        return false;
    }
}

console.log("Starting unit tests.");

// maxAmountOfStuffBits
check(BitSequence.maxAmountOfStuffBits(0) === 0);
check(BitSequence.maxAmountOfStuffBits(1) === 0);
check(BitSequence.maxAmountOfStuffBits(2) === 0);
check(BitSequence.maxAmountOfStuffBits(3) === 0);
check(BitSequence.maxAmountOfStuffBits(4) === 0);
check(BitSequence.maxAmountOfStuffBits(5) === 1);
check(BitSequence.maxAmountOfStuffBits(6) === 1);
check(BitSequence.maxAmountOfStuffBits(7) === 1);
check(BitSequence.maxAmountOfStuffBits(8) === 1);
check(BitSequence.maxAmountOfStuffBits(9) === 2);
check(BitSequence.maxAmountOfStuffBits(10) === 2);
check(BitSequence.maxAmountOfStuffBits(11) === 2);
check(BitSequence.maxAmountOfStuffBits(12) === 2);
check(BitSequence.maxAmountOfStuffBits(13) === 3);
check(BitSequence.maxAmountOfStuffBits(14) === 3);
check(BitSequence.maxAmountOfStuffBits(15) === 3);
check(BitSequence.maxAmountOfStuffBits(16) === 3);
check(BitSequence.maxAmountOfStuffBits(17) === 4);
check(BitSequence.maxAmountOfStuffBits(18) === 4);
check(BitSequence.maxAmountOfStuffBits(19) === 4);
check(BitSequence.maxAmountOfStuffBits(20) === 4);
check(BitSequence.maxAmountOfStuffBits(21) === 5);

// maxLengthAfterStuffing
check(BitSequence.maxLengthAfterStuffing(0) === 0);
check(BitSequence.maxLengthAfterStuffing(1) === 1);
check(BitSequence.maxLengthAfterStuffing(2) === 2);
check(BitSequence.maxLengthAfterStuffing(3) === 3);
check(BitSequence.maxLengthAfterStuffing(4) === 4);
check(BitSequence.maxLengthAfterStuffing(5) === 6);
check(BitSequence.maxLengthAfterStuffing(6) === 7);
check(BitSequence.maxLengthAfterStuffing(7) === 8);
check(BitSequence.maxLengthAfterStuffing(8) === 9);
check(BitSequence.maxLengthAfterStuffing(9) === 11);
check(BitSequence.maxLengthAfterStuffing(10) === 12);
check(BitSequence.maxLengthAfterStuffing(11) === 13);
check(BitSequence.maxLengthAfterStuffing(12) === 14);
check(BitSequence.maxLengthAfterStuffing(13) === 16);
check(BitSequence.maxLengthAfterStuffing(14) === 17);
check(BitSequence.maxLengthAfterStuffing(15) === 18);
check(BitSequence.maxLengthAfterStuffing(16) === 19);
check(BitSequence.maxLengthAfterStuffing(17) === 21);
check(BitSequence.maxLengthAfterStuffing(18) === 22);
check(BitSequence.maxLengthAfterStuffing(19) === 23);
check(BitSequence.maxLengthAfterStuffing(20) === 24);
check(BitSequence.maxLengthAfterStuffing(21) === 26);

// BitSequence constructor
let bits;
// Default
bits = new BitSequence();
check(arrayEqual(bits._sequence, []), bits._sequence);
check(bits._isStuffed === false);
// Binary string
bits = new BitSequence("    1 ")
check(arrayEqual(bits._sequence, [1]));
check(bits._isStuffed === false);
bits = new BitSequence("      0  1    ");
check(arrayEqual(bits._sequence, [0, 1]));
check(bits._isStuffed === false);
bits = new BitSequence("      0  1  101    ");
check(arrayEqual(bits._sequence, [0, 1, 1, 0, 1]));
check(bits._isStuffed === false);
bits = new BitSequence("      0  1  101    ", true);
check(arrayEqual(bits._sequence, [0, 1, 1, 0, 1]));
check(bits._isStuffed === true);
// Array of ints
bits = new BitSequence([0, 1, 0]);
check(arrayEqual(bits._sequence, [0, 1, 0]));
check(bits._isStuffed === false);
// Array of booleans
bits = new BitSequence([false, true, false]);
check(arrayEqual(bits._sequence, [0, 1, 0]));
check(bits._isStuffed === false);
// Hybrid array
bits = new BitSequence([1, false, "0"]);
check(arrayEqual(bits._sequence, [1, 0, 0]));
check(bits._isStuffed === false);
// Single bit integer
bits = new BitSequence(0);
check(arrayEqual(bits._sequence, [0]));
check(bits._isStuffed === false);
bits = new BitSequence(1);
check(arrayEqual(bits._sequence, [1]));
check(bits._isStuffed === false);
// Single bit boolean
bits = new BitSequence(false);
check(arrayEqual(bits._sequence, [0]));
check(bits._isStuffed === false);
bits = new BitSequence(true);
check(arrayEqual(bits._sequence, [1]));
check(bits._isStuffed === false);

// applyBitStuffing
check(new BitSequence().applyBitStuffing()
    .equal(new BitSequence("", true)));
check(new BitSequence("1").applyBitStuffing()
    .equal(new BitSequence("1", true)));
check(new BitSequence("1110101010").applyBitStuffing()
    .equal(new BitSequence("1110101010", true)));
check(new BitSequence("11111").applyBitStuffing()
    .equal(new BitSequence("111112", true)));
check(new BitSequence("111110000").applyBitStuffing()
    .equal(new BitSequence("11111200003", true)));

// Cannot double-stuff
try {
    new BitSequence("111110", true).applyBitStuffing();
    check(false, "Error nor raised");
} catch (TypeError) {
    // Error as expected
}

// length
check(new BitSequence().length() === 0);
check(new BitSequence("1").length() === 1);
check(new BitSequence("1110101010").length() === 10);
check(new BitSequence("11111").length() === 5);
check(new BitSequence("11111000001", true).length() === 11);

// toBinString
check(new BitSequence("").toBinString() === "");
check(new BitSequence("0").toBinString() === "0");
check(new BitSequence("10").toBinString() === "10");
check(new BitSequence("0010").toBinString() === "0010");
check(new BitSequence("0110101").toBinString() === "0110101");
check(new BitSequence("0110101").toBinString("[", "]") === "0110101");
check(new BitSequence("1111121", true).toBinString("[", "]") === "11111[0]1");
check(new BitSequence("0000031", true).toBinString("[", "]") === "00000[1]1");

// toBinStringWithSpacesRightAlign
check(new BitSequence("").toBinStringWithSpacesRightAlign() === "");
check(new BitSequence("0").toBinStringWithSpacesRightAlign() === "0");
check(new BitSequence("10").toBinStringWithSpacesRightAlign() === "10");
check(new BitSequence("0010").toBinStringWithSpacesRightAlign() === "0010");
check(new BitSequence("0110101").toBinStringWithSpacesRightAlign() === "011 0101");
check(new BitSequence("010110101").toBinStringWithSpacesRightAlign() === "0 1011 0101");
check(new BitSequence("010110101").toBinStringWithSpacesRightAlign("[", "]") === "0 1011 0101");
check(new BitSequence("0001111121", true).toBinStringWithSpacesRightAlign("[", "]") === "00 0111 11[0]1");
check(new BitSequence("0000031", true).toBinStringWithSpacesRightAlign("[", "]") === "000 00[1]1");

// toBinStringWithSpacesLeftAlign
check(new BitSequence("").toBinStringWithSpacesLeftAlign() === "");
check(new BitSequence("0").toBinStringWithSpacesLeftAlign() === "0");
check(new BitSequence("10").toBinStringWithSpacesLeftAlign() === "10");
check(new BitSequence("0010").toBinStringWithSpacesLeftAlign() === "0010");
check(new BitSequence("0110101").toBinStringWithSpacesLeftAlign() === "0110 101");
check(new BitSequence("010110101").toBinStringWithSpacesLeftAlign() === "0101 1010 1");
check(new BitSequence("010110101").toBinStringWithSpacesLeftAlign("[", "]") === "0101 1010 1");
check(new BitSequence("0001111121", true).toBinStringWithSpacesLeftAlign("[", "]") === "0001 1111 [0]1");
check(new BitSequence("0000031", true).toBinStringWithSpacesLeftAlign("[", "]") === "0000 0[1]1");

// toHexString
check(new BitSequence("").toHexString() === "");
check(new BitSequence("0").toHexString() === "0");
check(new BitSequence("00 0000 0000").toHexString() === "000");
check(new BitSequence("10").toHexString() === "2");
check(new BitSequence("0010").toHexString() === "2");
check(new BitSequence("0110101").toHexString() === "35");
check(new BitSequence("0 0001 1111").toHexString() === "01F");
check(new BitSequence("1 0001 1111").toHexString() === "11F");
check(new BitSequence("10 1111 1111").toHexString() === "2FF");

// Extend
check(new BitSequence("").extend("").toBinString() === "");
check(new BitSequence("").extend("0").toBinString() === "0");
check(new BitSequence("").extend("1").toBinString() === "1");
check(new BitSequence("").extend("01").toBinString() === "01");
check(new BitSequence("1").extend("00").toBinString() === "100");
check(new BitSequence("0").extend("11").toBinString() === "011");
check(new BitSequence("01").extend("00").toBinString() === "0100");
check(new BitSequence("01").extend(0).toBinString() === "010");
check(new BitSequence("01").extend(1).toBinString() === "011");
check(new BitSequence("01").extend(false).toBinString() === "010");
check(new BitSequence("01").extend(true).toBinString() === "011");

// Iterability
bits = new BitSequence("");
for (let b of bits) {
    check(false, "This should not be executed.");
}
bits = new BitSequence("1");
for (let b of bits) {
    check(b === 1);
}
bits = new BitSequence("101");
for (let b of bits) {
    check(b === 1 || b === 0);
}


// CRC
// The tested values in hex are: 0x0, 0xF1, 0x833, 0x34ec
// Test expected results are computed with
// https://www.ghsi.de/pages/subpages/Online%20CRC%20Calculation/index.php
// CRC 15 bits for classic CAN
check(crc15([0]) === 0);
check(crc15([1, 1, 1, 1, 0, 0, 0, 1]) === 0b110001011110110);
check(crc15([true, true, true, true, 0, 0, false, 1]) === 0b110001011110110);
check(crc15([1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1]) === 0b000010001110100);
check(crc15([1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0]) === 0b010001010101010);

// CRC 17 bits for CAN FD
check(crc17([0]) === 0);
check(crc17([1, 1, 1, 1, 0, 0, 0, 1]) === 0b00000001000111100);
check(crc17([1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1]) === 0b00100010111100111);
check(crc17([1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0])
    === 0b00000111010101101);

// CRC 21 bits for CAN FD
check(crc21([0]) === 0);
check(crc21([1, 1, 1, 1, 0, 0, 0, 1]) === 0b000010111011100111001);
check(crc21([1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1]) === 0b100110110111010101101);
check(crc21([1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0])
    === 0b001001011010101011100);


// CanFrame11Bit fields and whole frame
let frame;
let toComputeCrcOn;
let expectedWholeFrame;
let expectedWholeFrameStuffed;
let crcBits;

// Case 1: simple ID and empty payload
frame = new CanFrame11Bit(0, new Uint8Array(0));
check(frame.id === 0);
check(arrayEqual(frame.payload, new Uint8Array(0)));
check(frame.field01_startOfFrame().equal(new BitSequence(0)));
check(frame.field02_identifier().equal(new BitSequence("000 0000 0000")));
check(frame.field03_remoteTransmissionRequest().equal(new BitSequence(0)));
check(frame.field04_identifierExtensionBit().equal(new BitSequence(0)));
check(frame.field05_reservedBit().equal(new BitSequence(0)));
check(frame.field06_dataLengthCode().equal(new BitSequence("0000")));
check(frame.field07_dataField().equal(new BitSequence()));
toComputeCrcOn = new BitSequence("0  000 0000 0000  0  0  0000");
check(frame.field08_crc().equal(new BitSequence("000 0000 0000 0000")));
check(frame.field09_crcDelimiter().equal(new BitSequence(1)));
check(frame.field10_ackSlot().equal(new BitSequence(1)));
check(frame.field11_ackDelimiter().equal(new BitSequence(1)));
check(frame.field12_endOfFrame().equal(new BitSequence("111 1111")));
check(frame.field13_interFrameSpace().equal(new BitSequence("111")));
expectedWholeFrame = new BitSequence(
    "0  000 0000 0000  0  0  0  0000  " // Header
    + "" // Payload
    + "000 0000 0000 0000  " // CRC
    + "1  1  1  111 1111  111" // Rest of the trailer
);
check(frame.wholeFrame().equal(expectedWholeFrame));
expectedWholeFrameStuffed = new BitSequence(
    "0  000 03000 00300  0  0  03  0000  " // Header
    + "" // Payload
    + "0300 00030 0000 30000  " // CRC
    + "1  1  1  111 1111  111", // Rest of the trailer (unstuffed)
    true // This sequence is already stuffed
);
check(frame.wholeFrameStuffed().equal(expectedWholeFrameStuffed));

// Case 2: simple ID and payload with 1 byte
frame = new CanFrame11Bit(0, new Uint8Array(1));
check(frame.id === 0);
check(arrayEqual(frame.payload, new Uint8Array(1)));
check(frame.field01_startOfFrame().equal(new BitSequence(0)));
check(frame.field02_identifier().equal(new BitSequence("000 0000 0000")));
check(frame.field03_remoteTransmissionRequest().equal(new BitSequence(0)));
check(frame.field04_identifierExtensionBit().equal(new BitSequence(0)));
check(frame.field05_reservedBit().equal(new BitSequence(0)));
check(frame.field06_dataLengthCode().equal(new BitSequence("0001")));
check(frame.field07_dataField().equal(new BitSequence("0000 0000")));
toComputeCrcOn = new BitSequence("0  000 0000 0000  0  0  0001  0000 0000");
crcBits = new BitSequence(crc15(toComputeCrcOn).toString(2).padStart(15, "0"));
check(frame.field08_crc().equal(crcBits));
check(frame.field09_crcDelimiter().equal(new BitSequence(1)));
check(frame.field10_ackSlot().equal(new BitSequence(1)));
check(frame.field11_ackDelimiter().equal(new BitSequence(1)));
check(frame.field12_endOfFrame().equal(new BitSequence("111 1111")));
check(frame.field13_interFrameSpace().equal(new BitSequence("111")));
expectedWholeFrame = new BitSequence(
    "0  000 0000 0000  0  0  0  0001  " // Header
    + "0000 0000" // Payload
    + "100 0100 0010 0110" // CRC
    + "1  1  1  111 1111  111" // Rest of the trailer
);
check(frame.wholeFrame().equal(expectedWholeFrame));
expectedWholeFrameStuffed = new BitSequence(
    "0  000 03000 00300  0  0  03  0001  " // Header
    + "0000 03000" // Payload
    + "100 0100 0010 0110" // CRC
    + "1  1  1  111 1111  111", // Rest of the trailer (unstuffed)
    true // This sequence is already stuffed
);
check(frame.wholeFrameStuffed().equal(expectedWholeFrameStuffed));

// Case 3: complex ID and payload with 8 bytes
frame = new CanFrame11Bit(0x133, new Uint8Array(8));
frame.payload[0] = 0xFF;
frame.payload[7] = 0xFF;
check(frame.id === 0x133);
check(frame.payload[0] === 0xFF);
check(frame.payload[1] === 0);
check(frame.payload[6] === 0);
check(frame.payload[7] === 0xFF);
check(frame.field01_startOfFrame().equal(new BitSequence(0)));
check(frame.field02_identifier().equal(new BitSequence("001 0011 0011")));
check(frame.field03_remoteTransmissionRequest().equal(new BitSequence(0)));
check(frame.field04_identifierExtensionBit().equal(new BitSequence(0)));
check(frame.field05_reservedBit().equal(new BitSequence(0)));
check(frame.field06_dataLengthCode().equal(new BitSequence("1000")));
check(frame.field07_dataField().equal(new BitSequence(
    "1111 1111  0000 0000  0000 0000  0000 0000" +
    "0000 0000  0000 0000  0000 0000  1111 1111")));
toComputeCrcOn = new BitSequence(
    "0  001 0011 0011  0  0  0  1000  "
    + "1111 1111  0000 0000  0000 0000  0000 0000"
    + "0000 0000  0000 0000  0000 0000  1111 1111");
crcBits = new BitSequence(crc15(toComputeCrcOn).toString(2).padStart(15, "0"));
check(frame.field08_crc().equal(crcBits));
check(frame.field09_crcDelimiter().equal(new BitSequence(1)));
check(frame.field10_ackSlot().equal(new BitSequence(1)));
check(frame.field11_ackDelimiter().equal(new BitSequence(1)));
check(frame.field12_endOfFrame().equal(new BitSequence("111 1111")));
check(frame.field13_interFrameSpace().equal(new BitSequence("111")));
expectedWholeFrame = new BitSequence(
    "0  001 0011 0011  0  0  0  1000  " // Header
    + "1111 1111  0000 0000  0000 0000  0000 0000" // Payload
    + "0000 0000  0000 0000  0000 0000  1111 1111"  // Payload
    + "001 1000 0101 1111"  // CRC
    + "1  1  1  111 1111  111" // Rest of the trailer
);
check(frame.wholeFrame().equal(expectedWholeFrame));
expectedWholeFrameStuffed = new BitSequence(
    "0  001 0011 0011  0  0  0  1000  " // Header
    + "1111 12111  0000 03000  00300 00030  00003 0000" // Payload
    + "03000 00300  00030 00003  0000 03000  1111 12111"  // Payload
    + "001 1000 0101 11112" // CRC
    + "1  1  1  111 1111  111", // Rest of the trailer (unstuffed)
    true // This sequence is already stuffed
);
check(frame.wholeFrameStuffed().equal(expectedWholeFrameStuffed));


// CanFrame11Bit max frame length
// In the CanFrame11Bit class, the length is computed field by field. Here
// we compute the expected value manually. Note that the part after the CRC
// is not stuffed, so we must not include it in the computation of the amount
// of stuff bits.

// Header = SOF + ID + RTR + IDE + R0 + DLC
const headerLength = 1 + 11 + 1 + 1 + 1 + 4;
const crcLength = 15;
// Post CRC = CRC del + ACK slot + ACK del + EOF + pause
const postCrcLength = 1 + 1 + 1 + 7 + 3;

// Try it over all payload lengths
for (let i = 0; i <= 8; i++) {
    let payload = new Uint8Array(i);
    frame = new CanFrame11Bit(0, payload);
    let expectedMaxFrameLenAfterStuffing =
        BitSequence.maxLengthAfterStuffing(
            headerLength + payload.length * 8 + crcLength)
        + postCrcLength;
    check(frame.maxLengthAfterStuffing()
        === expectedMaxFrameLenAfterStuffing, payload);
}

// TODO tests for 29 bit ID frame
